File: python/src/simplex/phase1.py
from __future__ import annotations

import numpy as np
from scipy.linalg import lu_factor, lu_solve

from .logger import SimplexLogger
from .pivot import PivotRule, entering_index, leaving_index
from .preprocess import make_b_nonnegative
from .types import BasisState, LinearProgram, SolveStatus


def phase1(lp: LinearProgram, rule: PivotRule, logger: SimplexLogger,
           maxit: int = 1000, tol: float = 1e-8) -> tuple[BasisState, SolveStatus]:

    lp   = make_b_nonnegative(lp)
    m, n = lp.A.shape

    Aw = np.hstack([lp.A, np.eye(m)])
    cw = np.concatenate([np.zeros(n), -np.ones(m)])
    art_start = n  # 0-based: artificials are columns n..n+m-1

    basic    = list(range(n, n + m))
    nonbasic = list(range(n))

    logger.log_phase(1)

    for it in range(1, maxit + 1):
        B     = Aw[:, basic]
        N     = Aw[:, nonbasic]
        BF    = lu_factor(B)
        xb    = lu_solve(BF, lp.b)
        BinvN = lu_solve(BF, N)

        cb = cw[basic]
        cr = cw[nonbasic] - BinvN.T @ cb
        z  = float(cb @ xb)

        logger.log_iteration(it, BasisState(list(basic), list(nonbasic)), xb, z)

        j = entering_index(rule, cr, tol)

        if j is None:
            if z < -tol:
                return BasisState([], []), SolveStatus.INFEASIBLE

            fix_artificials_in_basis(Aw, basic, nonbasic, art_start)
            orig_basic    = [i for i in basic    if i < n]
            orig_nonbasic = [i for i in nonbasic if i < n]
            return BasisState(orig_basic, orig_nonbasic), SolveStatus.OPTIMAL

        d = BinvN[:, j]
        i = leaving_index(xb, d, BinvN, tol)

        if i is None:
            raise RuntimeError("Phase 1 is unexpectedly unbounded")

        basic[i], nonbasic[j] = nonbasic[j], basic[i]

    raise RuntimeError(f"Phase 1 did not converge in {maxit} iterations")


def fix_artificials_in_basis(Aw: np.ndarray, basic: list[int],
                              nonbasic: list[int], art_start: int) -> None:
    changed = True
    while changed:
        changed = False
        for pos, col in enumerate(basic):
            if col < art_start:
                continue

            BF     = lu_factor(Aw[:, basic])
            BinvAw = lu_solve(BF, Aw)
            row    = BinvAw[pos, :]

            swap_j = None
            for j, ncol in enumerate(nonbasic):
                if ncol < art_start and abs(row[ncol]) > 1e-10:
                    swap_j = j
                    break

            if swap_j is None:
                Aw[pos, :] = 0.0
            else:
                basic[pos], nonbasic[swap_j] = nonbasic[swap_j], basic[pos]
                changed = True
                break

File: go/simplex/phase1.go
package simplex

import "math"

// Phase1 finds an initial basic feasible solution using the two-phase method.
// It introduces artificial variables, solves the auxiliary LP, then removes
// any artificial variables that remain in the basis.
//
// Returns the resulting BasisState (in original variable indices) and a status.
// Status is Optimal if a BFS exists, Infeasible otherwise.
func Phase1(lp LinearProgram, rule PivotRule, logger SimplexLogger, maxit int, tol float64) (BasisState, SolveStatus) {
	lp = MakeBNonnegative(lp)
	m, n := lp.M, lp.N

	// Build augmented matrix Aw = [A | I_m], size m × (n+m), stored row-major.
	nw := n + m
	Aw := make([]float64, m*nw)
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			Aw[i*nw+j] = lp.A[i*n+j]
		}
		Aw[i*nw+n+i] = 1.0
	}

	// cw = [0...0, -1...-1], artificials have cost -1 (we maximize).
	cw := make([]float64, nw)
	for i := n; i < nw; i++ {
		cw[i] = -1.0
	}

	artStart := n // 0-based: artificials are columns n..n+m-1

	basic := make([]int, m)
	for i := range basic {
		basic[i] = n + i
	}
	nonbasic := make([]int, n)
	for i := range nonbasic {
		nonbasic[i] = i
	}

	logger.LogPhase(1)

	for it := 1; it <= maxit; it++ {
		nb := len(nonbasic)

		B := colsFromRowMajor(Aw, m, nw, basic)
		N := colsFromRowMajor(Aw, m, nw, nonbasic)

		lu := factorBasis(B)
		xb := solveVec(lu, lp.B)
		BinvN := solveMat(lu, N)
		binvNFlat := binvNToRowMajor(BinvN)

		cr := reducedCosts(gather(cw, nonbasic), BinvN, gather(cw, basic))
		z := dot(gather(cw, basic), xb)

		logger.LogIteration(it, BasisState{copyInts(basic), copyInts(nonbasic)}, xb, z)

		j, ok := rule.EnteringIndex(cr, tol)
		if !ok {
			// No positive reduced cost — optimal for auxiliary problem.
			if z < -tol {
				return BasisState{}, Infeasible
			}
			fixArtificialsInBasis(Aw, m, nw, basic, nonbasic, artStart)

			origBasic := filterLess(basic, n)
			origNonbasic := filterLess(nonbasic, n)
			return BasisState{origBasic, origNonbasic}, Optimal
		}

		d := denseCol(BinvN, j)
		i, ok := LeavingIndex(xb, d, binvNFlat, m, nb, tol)
		if !ok {
			panic("Phase 1 is unexpectedly unbounded")
		}

		basic[i], nonbasic[j] = nonbasic[j], basic[i]
	}

	panic("Phase 1 did not converge")
}

// fixArtificialsInBasis tries to pivot out any artificial variable still in the basis.
// If an artificial cannot be pivoted out (redundant row), that row of Aw is zeroed.
func fixArtificialsInBasis(Aw []float64, m, nw int, basic, nonbasic []int, artStart int) {
	changed := true
	for changed {
		changed = false
		for pos, col := range basic {
			if col < artStart {
				continue
			}
			B := colsFromRowMajor(Aw, m, nw, basic)
			lu := factorBasis(B)
			AwMat := matFromRowMajor(Aw, m, nw)
			BinvAw := solveMat(lu, AwMat)

			// Look for a non-artificial non-basic column with non-zero entry in this row.
			swapJ := -1
			for j, ncol := range nonbasic {
				if ncol < artStart && math.Abs(BinvAw.At(pos, ncol)) > 1e-10 {
					swapJ = j
					break
				}
			}

			if swapJ < 0 {
				// Redundant row — zero it out so it doesn't affect future factorizations.
				for j := 0; j < nw; j++ {
					Aw[pos*nw+j] = 0.0
				}
			} else {
				basic[pos], nonbasic[swapJ] = nonbasic[swapJ], basic[pos]
				changed = true
				break
			}
		}
	}
}

// copyInts returns a copy of a slice of ints.
func copyInts(s []int) []int {
	c := make([]int, len(s))
	copy(c, s)
	return c
}

// filterLess returns elements of s that are strictly less than threshold.
func filterLess(s []int, threshold int) []int {
	out := make([]int, 0, len(s))
	for _, v := range s {
		if v < threshold {
			out = append(out, v)
		}
	}
	return out
}
